src/lib/rule2F.ts

```diff
@@ -0,0 +1,207 @@
+ */
+function allowsNameFromContent(elem: HTMLElement): boolean {
+  const elemRole: string =
+    elem.getAttribute('role')?.trim().toLowerCase() ?? '';
+  if (nameFromContentRoles.includes(elemRole)) {
+    return true;
+  }
+
+  const elemNodeName: string = elem.nodeName.toLowerCase();
+  if (nameFromContentElemNodeName.includes(elemNodeName)) {
+    return true;
+  }
+
+  const nameFromContentFunction = nameFromContentFunctionOfElem[elemNodeName];
+  if (nameFromContentFunction) {
+    return nameFromContentFunction(elem);
+  }
+
+  return false;
+}
+
+export {allowsNameFromContent as allowsNameFromContent_TEST};
+
+/**
+ * Checks if 'elem' in with 'context' satisfies the conditions
+ * necessary to apply rule 2F.
+ * @param elem - elem whose text alternative is being computed
+ * @param context - additional information about the context of elem
+ * @return - whether or not rule 2Fs condition has been satisfied
+ */
+function rule2FCondition(elem: HTMLElement, context: Context): boolean {
+  if (allowsNameFromContent(elem)) {
+    return true;
+  }
+
+  if (context.ariaLabelledbyReference) {
+    return true;
+  }
+
+  if (context.isLabelReference) {
+    context.inherited.isLabelDescendant = true;
+    return true;
+  }
+
+  if (context.inherited.isLabelDescendant) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Implementation of rule 2F
+ * @param node - node whose text alternative is being calculated
+ * @param context - additional info relevant to the calculation of nodes
+ * text alternative
+ * @return - text alternative for node if the conditions of 2F are satisfied,
+ * null otherwise.
+ */
+export function rule2F(node: Node, context: Context): string | null {
+  if (!(node instanceof HTMLElement)) {
+    return null;
+  }
+
+  if (!rule2FCondition(node, context)) {
+    return null;
+  }
+
+  // CSS generated text seems to have value 'on' if it doesn't exist. (rather than null?)
+  // Not sure if this is a bug or I've done something wrong.
+  let cssBeforeContent: string = window
+    .getComputedStyle(node, ':before')
+    .content.slice(1, -1);
+  if (cssBeforeContent === 'on') {
+    cssBeforeContent = '';
+  }
+  let cssAfterContent: string = window
+    .getComputedStyle(node, ':after')
+    .content.slice(1, -1);
+  if (cssAfterContent === 'on') {
+    cssAfterContent = '';
+  }
+
+  const textAlterantives: string[] = [];
+  node.childNodes.forEach(childNode => {
+    if (!context.inherited.visitedNodes.includes(childNode)) {
+      context.inherited.visitedNodes.push(childNode);
+      const textAlterantive = computeTextAlternative(
+        childNode,
+        resetUninherited(context)
+      );
+      // Only consider non-empty text alternatives to avoid multiple spaces
+      // between words from .join(' ')
+      if (textAlterantive !== '') {
+        textAlterantives.push(textAlterantive);
+      }
+    }
+  });
+
+  const accumulatedText = textAlterantives.join(' ');
```

Comment:
AFAICT the spec doesn't mention adding a space between each one? It does seem like the obviously right thing to do though

Reply:
Yeah this is an interesting question, I joined with a space because that's what Chrome seems to do but it doesn't mention to do so in the spec, as you said. This might be something the spec should include as the words won't necessarily make sense if they're concatenated without a space in between.